riscv/proc_trace.cc

```diff
@@ -0,0 +1,325 @@
+        tval &= 0xffffffff;
+    }
+    fprintf(m_itrace_fd, "1,%lx,%lx,%x,%x,%lx,%lx,%x\n",
+            addr, m_insn_binary, m_prv, m_ex, m_ex_cause, tval, m_interrupt);
+    fflush(m_itrace_fd);
+
+    m_ex = 0;
+    m_ex_cause = 0;
+    m_tval = 0;
+    m_interrupt = 0;
+}
+
+void proc_trace_t::set_addr(reg_t addr) {
+    m_addr = addr;
+}
+
+void proc_trace_t::set_insn(insn_t insn, uint64_t binary) {
+    m_has_insn = true;
+    m_insn = insn;
+    m_insn_binary = binary;
+    if (m_dtrace_fd) {
+        m_dtype = DTYPE_INVALID;
+        m_recorded_dtype = DTYPE_INVALID;
+        m_csr_which = CSR_INVALID;
+        m_csr_write = false;
+        m_csr_load_data = 0;
+        m_csr_size = 0;
+
+        bool is_compact = (binary & 0x3) != 0x3;
+        m_is_amo = false;
+        if (is_compact) {
+            // Not interested in load immediate
+            uint64_t inst_15_13 = (binary >> 13) & ((insn_bits_t(1) << 3) - 1);
+            uint64_t inst_1_0 = binary & 0x3;
+            m_is_load = (((inst_1_0 == 0) || (inst_1_0 == 2)) && (inst_15_13 > 0) && (inst_15_13 < 4));
+            m_is_store = ((inst_1_0 == 0) || (inst_1_0 == 2)) & (inst_15_13 > 4);
+        } else {
+            uint64_t inst_6_0 = binary & ((insn_bits_t(1) << 7) - 1);
+            uint64_t inst_4_2 = binary >> 2 & ((insn_bits_t(1) << 3) - 1);
+            uint64_t inst_6_5 = binary >> 5 & ((insn_bits_t(1) << 2) - 1);
+            m_is_load = (inst_4_2 < 2) && (inst_6_5 == 0);
+            m_is_store = (inst_4_2 < 2) && (inst_6_5 == 1);
+            m_is_amo = (inst_6_0 == 0x2f);
+        }
+
+        if (m_is_load)
+            set_dtype(DTYPE_LOAD);
+        if (m_is_store)
+            set_dtype(DTYPE_STORE);
+        if (m_data_trace_debug)
+            fprintf(m_dtrace_fd, "DEBUG: INSN@%lx: %lx\n", m_addr, m_insn_binary);
+    }
+}
+
+void proc_trace_t::set_priv(uint8_t prv) {
+    m_prv = prv;
+}
+
+void proc_trace_t::set_exception(reg_t cause) {
+    m_ex = 1;
+    m_ex_cause = cause;
+}
+
+void proc_trace_t::set_tval(reg_t tval) {
+    m_tval = tval;
+}
+
+void proc_trace_t::set_interrupt(uint8_t interrupt) {
+    m_interrupt = interrupt;
+}
+
+void proc_trace_t::set_is_32bit_isa(bool is_32bit_isa) {
+    m_is_32bit_isa = is_32bit_isa;
+}
+
+void proc_trace_t::record_load(reg_t addr, uint64_t res, size_t size, data_src_t src) {
+    if (m_dtrace_fd) {
+        if (!m_has_insn || m_is_amo)
+            return;
+
+        if (!m_is_load) {
+            // It's not clear why these are being triggered.
+            // They seem to come from DATA_SRC_LOAD_TLB_1
+            if (m_data_trace_debug)
+                fprintf(m_dtrace_fd, "NOT LOAD SRC %d %lx ld %d st %d amo %d %lx\n",
+                        src, addr, m_is_load, m_is_store, m_is_amo, m_insn_binary);
+            else
+                fprintf(stderr, "NOT LOAD SRC %d %lx ld %d st %d amo %d %lx\n",
+                        src, addr, m_is_load, m_is_store, m_is_amo, m_insn_binary);
```

Comment:
No need to duplicate the entire` fprintf()` when the only difference is the output filehandle

Reply:
Changed to using C++ streams.